File: FlutterPlugin/ios/Classes/LineNative/DaroLineNativeAdPlatformView.swift
import Flutter
import Daro

class DaroLineNativeAdPlatformView: NSObject, FlutterPlatformView {
    private let containerView: UIView
    private var lineNativeAdView: DaroAdLineBannerView
    private let adId: Int
    private let channel: FlutterMethodChannel

    init(
        frame: CGRect,
        adId: Int,
        adUnitId: String,
        configuration: DaroLineNativeAdConfiguration,
        channel: FlutterMethodChannel
    ) {
        self.containerView = UIView(frame: frame)
        self.containerView.backgroundColor = .clear
        self.adId = adId
        self.channel = channel

        let adUnit = DaroAdUnit(unitId: adUnitId)
        self.lineNativeAdView = DaroAdLineBannerView(unit: adUnit, autoLoad: false)
        self.lineNativeAdView.configuration = configuration

        super.init()

        setupLineNativeAd()
    }

    func view() -> UIView {
        return containerView
    }

    private func setupLineNativeAd() {
        lineNativeAdView.listener.onAdLoadSuccess = { [weak self] (_: DaroAdUnit, _: DaroAdInfo?) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdLoaded")
        }

        lineNativeAdView.listener.onAdLoadFail = { [weak self] (error: Error) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdFailedToLoad", error: ["code": -1, "message": error.localizedDescription])
        }

        lineNativeAdView.listener.onAdClicked = { [weak self] (_: DaroAdInfo?) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdClicked")
        }

        lineNativeAdView.listener.onAdImpression = { [weak self] (_: DaroAdInfo?) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdImpression")
        }

        containerView.addSubview(lineNativeAdView)
        lineNativeAdView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            lineNativeAdView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            lineNativeAdView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
            lineNativeAdView.topAnchor.constraint(equalTo: containerView.topAnchor),
            lineNativeAdView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
        ])

        lineNativeAdView.loadAd()
    }

    private func sendEvent(event: String, error: [String: Any]? = nil) {
        var arguments: [String: Any] = [
            "adId": adId,
            "event": event,
        ]

        if let error = error {
            arguments["error"] = error
        }

        channel.invokeMethod("onAdEvent", arguments: arguments)
    }
}

File: FlutterPlugin/android/src/main/kotlin/so/daro/flutter/linenative/DaroLineNativeAdPlatformView.kt
package so.daro.flutter.linenative

import android.content.Context
import android.graphics.Color
import android.view.View
import android.widget.FrameLayout
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import droom.daro.core.adunit.DaroNativeAdUnit
import droom.daro.core.model.DaroAdInfo
import droom.daro.core.model.DaroAdLoadError
import droom.daro.core.model.DaroNativeAdBinder
import droom.daro.core.model.DaroViewAd
import droom.daro.core.model.nativetemplate.DaroNativeAdTemplate
import droom.daro.view.DaroAdViewListener
import droom.daro.view.DaroNativeAdView

class DaroLineNativeAdPlatformView(
    context: Context,
    private val adId: Int,
    private val adUnitId: String,
    private val backgroundColor: Int?,
    private val contentColor: Int?,
    private val adMarkLabelTextColor: Int?,
    private val adMarkLabelBackgroundColor: Int?,
    private val channel: MethodChannel
) : PlatformView {

    private val containerView: FrameLayout = FrameLayout(context)
    private var nativeAdView: DaroNativeAdView? = null

    init {
        setupLineNativeAd(context)
    }

    override fun getView(): View = containerView

    override fun dispose() {
        nativeAdView?.destroy()
        nativeAdView = null
    }

    private fun setupLineNativeAd(context: Context) {
        val adUnit = DaroNativeAdUnit(
            key = adUnitId,
            placement = ""
        )
        val nativeAdView = DaroNativeAdView(context, adUnit)
        this.nativeAdView = nativeAdView

        val template = DaroNativeAdTemplate.LineCenter(
            backgroundColor = backgroundColor ?: Color.TRANSPARENT,
            contentColor = contentColor ?: Color.BLACK,
            adMarkLabelBackgroundColor = adMarkLabelBackgroundColor ?: Color.LTGRAY,
            adMarkLabelTextColor = adMarkLabelTextColor ?: Color.WHITE
        )
        val adBinder = DaroNativeAdBinder.fromTemplate(context, template)
        nativeAdView.setAdBinder(adBinder)

        nativeAdView.setListener(object : DaroAdViewListener {
            override fun onAdLoadSuccess(ad: DaroViewAd, adInfo: DaroAdInfo) {
                sendEvent("onAdLoaded")
            }

            override fun onAdLoadFail(err: DaroAdLoadError) {
                sendEvent("onAdFailedToLoad", mapOf(
                    "code" to (err.code ?: -1),
                    "message" to (err.message ?: "Unknown error")
                ))
            }

            override fun onAdClicked(adInfo: DaroAdInfo) {
                sendEvent("onAdClicked")
            }

            override fun onAdImpression(adInfo: DaroAdInfo) {
                sendEvent("onAdImpression")
            }
        })

        val layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
        containerView.addView(nativeAdView, layoutParams)

        nativeAdView.loadAd()
    }

    private fun sendEvent(event: String, error: Map<String, Any>? = null) {
        val arguments = mutableMapOf<String, Any>(
            "adId" to adId,
            "event" to event
        )
        error?.let { arguments["error"] = it }

        channel.invokeMethod("onAdEvent", arguments)
    }
}
